File: game.py
import numpy as np

from copy import deepcopy


def pretty_print_board(board, probs=None):
    mp = {
        1: '\033[92mX\033[0m',
        -1: '\033[94mO\033[0m',
        0: ' '
    }

    n, m = board.shape
    b = [
        [mp[board[i, j]] for j in range(m)] for i in range(n)
    ]

    if probs is not None:
        for i, row in enumerate(b):
            for j, x in enumerate(row):
                if x == ' ':
                    b[i][j] = f"{probs[i * m + j]:.3f}"
    cell_w = 7
    print("-"*(cell_w*m +m))
    for row in b:
        row_str = "|"
        row_str += (" " * cell_w + '|') * m
        print(row_str)
        x_row = "|"
        for x in row:
            if x in "XO ":
                x_row += f"   {x}   |"
            else:
                x_row += f" {x} |"
        print(x_row)
        print(row_str)
        print("-"*(cell_w*m +m))


class GameState:
    WIN = 0
    TIE = 1
    IN_PROGRESS = 2

class TicTacToe():
    def __init__(self, N, M, win_req):
        self.win_req = win_req
        self.size = (N, M)
        self.board = np.zeros(self.size)

        self.winner = None
        self.turn = 1
        self.history = []

    def _map_moves(self, moves):
        a = np.zeros(self.size[0] * self.size[1])
        for i, j in moves:
            a[i*self.size[1] + j] = 1
        return a

    def allowed_moves(self):
        moves =  [
            (i, j) for i, j in zip(*(self.board == 0).nonzero())
        ]
        return self._map_moves(moves)

    def state(self):
        return deepcopy(self.board)

    def move(self, k):
        i = k // self.size[1]
        j = k % self.size[1]
        if self.board[i, j] != 0:
            raise ValueError(f"Invalid move [{i}][{j}] already taken")

        self.history.append((self.turn, (i, j)))
        self.board[i, j] = self.turn

        self.turn *= -1
        return self._check_winner()

    def _check_winner(self):
        player, last_move = self.history[-1]

        i, j = last_move
        in_row = 0

        # check horizontal and vertical
        end = j
        while end < self.size[1] and self.board[i, end] == player:
            end += 1
        start = j
        while start >= 0 and self.board[i, start] == player:
            start -= 1
        in_row = max(end - start - 1, in_row)

        end = i
        while end < self.size[0] and self.board[end, j] == player:
            end += 1
        start = i
        while start >= 0 and self.board[start, j] == player:
            start -= 1
        in_row = max(end - start - 1, in_row)

        # diagonals
        end_i = i
        end_j = j
        while end_i < self.size[0] and end_j < self.size[1] and self.board[end_i, end_j] == player:
            end_i += 1
            end_j += 1

        start_i = i
        start_j = j
        while start_i >= 0 and start_j >= 0 and self.board[start_i, start_j] == player:
            start_i -= 1
            start_j -= 1
        in_row = max(end_i - start_i - 1, in_row)

        end_i = i
        end_j = j
        while end_i < self.size[0] and end_j >= 0 and self.board[end_i, end_j] == player:
            end_i += 1
            end_j -= 1

        start_i = i
        start_j = j
        while start_i >= 0 and start_j < self.size[1] and self.board[start_i, start_j] == player:
            start_i -= 1
            start_j += 1
        in_row = max(end_i - start_i - 1, in_row)

        if in_row >= self.win_req:
            self.winner = player
            return GameState.WIN

        if len(self.allowed_moves()) == 0:
            self.winner = 0
            return GameState.TIE

        return GameState.IN_PROGRESS

    def in_progress(self):
        return self.winner is None

File: cpp/game.cpp
#include "game.hpp"

#include <stdexcept>
#include <sstream>
#include <numeric>

TicTacToe::TicTacToe(int n, int m, int req):
    n(n), m(m), req(req), turn(1), board(n, std::vector<int>(m, 0)), _allowed_moves(m*n, 1){
  winner = 2;
}

std::vector<float> TicTacToe::allowed_moves() {
  return _allowed_moves;
}

std::vector<float> TicTacToe::state() const {
  std::vector<float> res(n*m*3, 0.);
  for (int i = 0; i < n; i++) {
    for (int j = 0; j< m; j++) {
      for (int t = 0; t < 3; t++) {
        int x = board[i][j];
        if ((t == 0 and x == 1) or (t == 1 and x == -1)) {
          res[3 *(i * m + j) + t] = 1;
        } else if (t == 2) {
          res[3 * (i * m + j) + t] = (1 - turn) / 2;
        }
      }
    }
  }
  return res;
}

int TicTacToe::move(int k) {
  int i = k / m;
  int j = k % m;
  if (board[i][j] != 0) {
    std::ostringstream oss;
    oss << "move [" << i << "][" << j << "] taken by " << turn << "\n" << *this;
    throw std::invalid_argument(oss.str());
  }

  board[i][j] = turn;
  _allowed_moves[k] = 0;
  turn *= -1;
  last_move = k;

  return check_winner();
}


int TicTacToe::check_winner() {
  int i = last_move / m;
  int j = last_move % m;

  int player = turn * -1;
  int in_row = 0;

  int end = j;
  while (end < m and board[i][end] == player) {
    end += 1;
  }
  int start = j;
  while (start >= 0 and board[i][start] == player) {
      start -= 1;
  }
  in_row = std::max(end - start - 1, in_row);

  end = i;
  while (end < n and board[end][j] == player) {
    end += 1;
  }
  start = i;
  while (start >= 0 and board[start][j] == player) {
      start -= 1;
  }
  in_row = std::max(end - start - 1, in_row);

  int end_i = i;
  int end_j = j;
  while (end_i < n and end_j < m and board[end_i][end_j] == player) {
    end_i += 1;
    end_j += 1;
  }

  int start_i = i;
  int start_j = j;
  while (start_i >= 0 and start_j >= 0 and board[start_i][start_j] == player) {
    start_i -= 1;
    start_j -= 1;
  }
  in_row = std::max(end_i - start_i - 1, in_row);

  end_i = i;
  end_j = j;
  while (end_i < n and end_j >= 0 and board[end_i][end_j] == player) {
    end_i += 1;
    end_j -= 1;
  }

  start_i = i;
  start_j = j;
  while (start_i >= 0 and start_j < m and board[start_i][start_j] == player) {
    start_i -= 1;
    start_j += 1;
  }
  in_row = std::max(end_i - start_i - 1, in_row);

  if (in_row >= req) {
    winner = player;
    return player;
  }

  if (std::accumulate(_allowed_moves.begin(), _allowed_moves.end(), 0) == 0) {
    winner = 0;
    return winner;
  }

  winner = 2;
  return 2;
}

int TicTacToe::get_winner() {
  return winner;
}

std::ostream& operator<<(std::ostream& os, const TicTacToe& t) {
  os << std::string(t.m * 7 + t.m, '-') << std::endl;
  std::string emp = "|";
  for (int i = 0; i < t.m; i++) emp += "       |";
  for (int i = 0; i < t.n; i ++) {
    os << emp << std::endl;
    os << "|";
    for (int j = 0; j < t.m; j++) {
      os << "   ";
      if (t.board[i][j] == 1) {
        os << "\033[92mX\033[0m";
      } else if (t.board[i][j] == -1) {
        os << "\033[94mO\033[0m";
      } else {
        os << " ";
      }
      os << "   |";
    }

    os << std::endl << emp << std::endl;
    os << std::string(t.m * 7 + t.m, '-') << std::endl;
  }
  return os;
}
